rtt/TaskContext.cpp

```diff
@@ -344,15 +344,26 @@ namespace RTT
     {
         if (this->isRunning())
             return false;
-        if ( new_act == 0) {
+
+        // refuse to setActivity from our own active thread
+        if (our_act) {
+            if (our_act->isActive() && our_act->thread() && our_act->thread()->isSelf()) {
+                log(Error) << "Cannot set the activity of TaskContext "
+                           << this->getName() << " from its own thread." << endlog();
+                delete new_act;
```

Comment:
`our_act` is a shared_ptr. I'm not really sure what are supposed to be the ownership rules here.

In any case, whatever you believe the rules are, please document `setActivity` accordingly. Including and especially about this new change.

Reply:
You are right. The intended behavior is described in [TaskContext.hpp:130-143](https://github.com/orocos-toolchain/rtt/blob/a5a8fd73d5ee3522cc7a65d0167787f5218acae9/rtt/TaskContext.hpp#L130-L143):
```cpp
        /**
         * Sets the activity of this TaskContext. The
         * activity is owned by the TaskContext and you should
         * not hold a pointer to it after this call. Use
         * getActivity() lateron to retrieve a safe pointer to it.
         * @param new_act The new activity for this TaskContext,
         * which becomes owned by this TaskContext, in case this method
         * returns true.
         * @return false if this->isRunning(). You can not change the
         * activity of a TaskContext once it is running. In that case,
         * new_act is not destroyed.
         * @note This function may not be called from the current
         * ExecutionEngine thread (OwnThread), another thread (ClientThread) must call this function.
         */
```

So in case of an error (like here) the method should not destroy the activity and it is up to the caller to check the return value and eventually delete the instance.

I will remove the `delete new_act` to comply with the documentation of `setActivity()`.

Reply:
Done in https://github.com/orocos-toolchain/rtt/pull/171/commits/0e117bd6675f51776f4cc8d36bfa43e84348dbba.